### PR TITLE
Use SampleMPs in dynamic_one_shot

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
The native MCM workflow breaks the device API where a sequence of MeasurementProcess objects is expected in the output.

**Description of the Change:**
Introduce SampleMPs in the auxiliary tape. Pass the mid_measurements dictionary around simulate and sampling to return the correct sample measurements. Modify the dynamic_one_shot transform post-processing function accordingly. 

**Benefits:**
Conform to current API.
Road to jax.jit support.

**Possible Drawbacks:**
Ad hoc post-processing required in measure_with_samples.

**Related GitHub Issues:**
[sc-60945]

### CODE REVIEW

File: pennylane/devices/qubit/apply_operation.py [hunk 1]
```diff
@@ -261,21 +261,14 @@ def apply_mid_measure(
...
-
-    try:  # pragma: no cover
-        sample = np.random.binomial(1, probs[1])
-    except ValueError as e:  # pragma: no cover
-        if probs[1] > 1:  # MachEps error, safe to catch
-            sample = np.random.binomial(1, np.round(probs[1], 15))
-        else:  # Other general error, continue to fail
-            raise e
-
+    sample = np.random.binomial(1, probs[1])
```
Comment (albi3ro) [inline pennylane/devices/qubit/apply_operation.py:268]: Have we fixed the mach eps error with this change somehow?

File: pennylane/devices/qubit/sampling.py [hunk 2]
```diff
@@ -217,6 +225,12 @@ def measure_with_samples(
         res for _, res in sorted(list(enumerate(all_res)), key=lambda r: flat_indices[r[0]])
     )
 
+    # append MCM samples
+    if mid_measurements:
+        sorted_res = list(sorted_res)
+        sorted_res.extend(list(mid_measurements.values()))
+        sorted_res = tuple(sorted_res)
```
Comment (albi3ro) [inline pennylane/devices/qubit/sampling.py:232]:
```suggestion
        sorted_res += tuple(mid_measurements.values())
```
Fewer type conversions.

File: pennylane/tape/tape.py [hunk 3]
```diff
@@ -40,29 +46,40 @@ def _err_msg_for_some_meas_not_qwc(measurements):
     )
 
 
-def _validate_computational_basis_sampling(measurements):
+def _validate_computational_basis_sampling(tape):
```
Comment (albi3ro) [inline pennylane/tape/tape.py:49]: I'm a little curious about what forcing the change in this code.  If it's a mid-circuit measurement, shouldn't it be completely independent of what basis we are measuring in at the end?

File: pennylane/transforms/dynamic_one_shot.py [hunk 4]
```diff
@@ -101,13 +103,40 @@ def processing_fn(results, has_partitioned_shots=None):
                 final_results.append(processing_fn(results[0:s], has_partitioned_shots=False))
                 del results[0:s]
             return tuple(final_results)
+        all_mcms = []
+        for op in aux_tape.operations:
+            if isinstance(op, MidMeasureMP):
+                all_mcms.append(op)
```
Comment (albi3ro) [inline pennylane/transforms/dynamic_one_shot.py:109]:
```suggestion
        all_mcms = [op for op in aux_tape.operations if isinstance(op, MidMeasureMP)]
```

File: pennylane/transforms/dynamic_one_shot.py [hunk 5]
```diff
@@ -101,13 +103,40 @@ def processing_fn(results, has_partitioned_shots=None):
...
+        n_mcms = len(all_mcms)
+        post_process_tape = qml.tape.QuantumScript(
+            aux_tape.operations,
+            aux_tape.measurements[0:-n_mcms],
+            shots=aux_tape.shots,
+            trainable_params=aux_tape.trainable_params,
+        )
+        sca_results = len(post_process_tape.measurements) == 0 and len(aux_tape.measurements) == 1
```
Comment (albi3ro) [inline pennylane/transforms/dynamic_one_shot.py:117]: What's `sca` stand for?

File: pennylane/transforms/dynamic_one_shot.py [hunk 6]
```diff
@@ -101,13 +103,40 @@ def processing_fn(results, has_partitioned_shots=None):
...
+        results = list(compress(results, mcm_mask))
         all_shot_meas, list_mcm_values_dict, valid_shots = None, [], 0
-        for res in results:
-            one_shot_meas, mcm_values_dict = res
-            if one_shot_meas is None:
-                continue
+        for i, res in enumerate(results):
+            samples = (
+                [res]
+                if len(post_process_tape.measurements) == 0 and len(aux_tape.measurements) == 1
```
Comment (albi3ro) [inline pennylane/transforms/dynamic_one_shot.py:128]: If this the same `sca_results` condition?

File: pennylane/devices/qubit/apply_operation.py (hunk 1 above)
Comment (vincentmr) [reply pennylane/devices/qubit/apply_operation.py]: I was not aware of this change. I'm not sure how it got deleted TBH, maybe a `git merge` + conflict resolution mistake.

File: pennylane/tape/tape.py (hunk 3 above)
Comment (vincentmr) [reply pennylane/tape/tape.py:49]: I agree, but without this change, the (old API) L-Kokkos device raises an error
```
pennylane.QuantumFunctionError: Only observables that are qubit-wise commuting Pauli words can be returned on the same wire, some of the following measurements do not commute:
```
The current resolution feels hacky, but I could not find a sleeker way to resolve this.

File: pennylane/transforms/dynamic_one_shot.py (hunk 6 above)
Comment (vincentmr) [reply pennylane/transforms/dynamic_one_shot.py]: Yes.

File: pennylane/devices/qubit/simulate.py [hunk 7]
```diff
@@ -170,15 +170,19 @@ def measure_final_state(circuit, state, is_state_batched, rng=None, prng_key=Non
...
     """
 
     circuit = circuit.map_to_standard_wires()
 
+    # analytic case
+
     if not circuit.shots:
-        # analytic case
+        if mid_measurements is not None:
+            raise TypeError("Native mid-circuit measurements are not supported with finite shots.")
```
Comment (albi3ro) [inline pennylane/devices/qubit/simulate.py:185]:
```suggestion
            raise TypeError("Native mid-circuit measurements are not supported without finite shots.")
```

File: pennylane/tape/tape.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/tape/tape.py:49]: weird but ok 🤔

File: pennylane/devices/qubit/simulate.py (hunk 7 above)
Comment (albi3ro) [reply pennylane/devices/qubit/simulate.py:185]: and needs a test.

**Review by albi3ro — COMMENTED** [review]
All the source code looks good to me. 

I'm surprised no tests failed with this change.  Could we perhaps add some more unit tests to `tests/transforms/test_dynamic_one_shot.py` to assert the output batch when applied to a tape looks correct?

**Review by mudit2812 — APPROVED** [review]
Looks good to me. Approving since the only thing I can see that needs to be changed is the additional tests Christina suggested.

**Comment by vincentmr** [comment]
> Looks good to me. Approving since the only thing I can see that needs to be changed is the additional tests Christina suggested.

Thanks @mudit2812 . I added some tests in `tests/transforms/test_dynamic_one_shot.py` as suggested by @albi3ro . Do you think there is enough there?

**Comment by mudit2812** [comment]
The added tests seem sound to me. Could you also add a test to check the correctness for broadcasted tapes? The number of tapes in that case should be `batch_size * n_shots`

**Pull request merged**